### PR TITLE
firstboot: Add --root-shell option and tighten up passwd/shadow handling

### PR DESCRIPTION
**Pull request by daandemeyer**
The first commit adds --root-shell and related options to configure the shell for the root user.

Second commit tries to tighten up the whole passwd/shadow story a bit more. There are a lot of edge cases when we're modifying both files since they generally should be modified in lockstep so I tried to reduce the number of possibles scenarios a bit by either modifying neither or modifying both but nothing inbetween that.

I also added some logic so that if we have to add the root entries to passwd and shadow ourselves and no root password is provided by the user, we lock the root account so as to not cause security issues by accident.

Tested manually and via my modified mkosi that uses systemd-firstboot. Maybe we should eventually add a firstboot test to make it a bit harder to accidentally break stuff.

Ideally this could still go in for v246 since it makes stuff a bit more strict than before (if we want this in the first place of course).

### CODE REVIEW

File: src/firstboot/firstboot.c [hunk 1]
```diff
@@ -748,13 +788,35 @@ static int process_root_password(void) {
         if (lock < 0)
                 return log_error_errno(lock, "Failed to take a lock: %m");
 
-        if (arg_delete_root_password) {
+        if (arg_copy_root_shell && arg_root) {
+                struct passwd *p;
+
+                p = getpwnam("root");
```
Comment (keszybz) [inline src/firstboot/firstboot.c:803]: `errno = 0` here first.

File: src/firstboot/firstboot.c [hunk 2]
```diff
@@ -748,13 +788,35 @@ static int process_root_password(void) {
...
+                if (!p && errno != ENOENT) {
+                        if (!errno)
+                                errno = EIO;
+
+                        return log_error_errno(errno, "Failed to find passwd entry for root: %m");
```
Comment (keszybz) [inline src/firstboot/firstboot.c]: `return log_error_errno(errno_or_else(EIO), "Failed to find passwd entry for root: %m");`

File: src/firstboot/firstboot.c [hunk 3]
```diff
@@ -748,13 +788,35 @@ static int process_root_password(void) {
         if (lock < 0)
                 return log_error_errno(lock, "Failed to take a lock: %m");
 
-        if (arg_delete_root_password) {
+        if (arg_copy_root_shell && arg_root) {
+                struct passwd *p;
+
+                p = getpwnam("root");
+                if (!p && errno != ENOENT) {
```
Comment (keszybz) [inline src/firstboot/firstboot.c]: That seems iffy. getpwnam returns NULL if the name is not found, not ENOENT.

File: src/firstboot/firstboot.c [hunk 4]
```diff
@@ -638,7 +679,7 @@ static int write_root_passwd(const char *passwd_path, const char *password) {
                         .pw_gid = 0,
                         .pw_gecos = (char *) "Super User",
                         .pw_dir = (char *) "/root",
-                        .pw_shell = (char *) "/bin/sh",
+                        .pw_shell = shell ? (char *) shell : (char *) "/bin/sh",
```
Comment (keszybz) [inline src/firstboot/firstboot.c]: `(char*) (shell ?: "/bin/sh")` ?

Comment (daandemeyer) [reply src/firstboot/firstboot.c]: Nice one! I keep forgetting about all this GCC goodness.

File: src/firstboot/firstboot.c (hunk 3 above)
Comment (daandemeyer) [reply src/firstboot/firstboot.c]: So I changed this to something completely different in the second commit, same with the EIO comment. Should I spend the time to fix up the first commit or should I just merge them? They're related enough for me to justify merging them.

File: src/firstboot/firstboot.c [hunk 5]
```diff
@@ -1108,6 +1177,13 @@ static int parse_argv(int argc, char *argv[]) {
                         arg_root_password_is_hashed = true;
                         break;
 
+                case ARG_ROOT_SHELL:
+                        r = free_and_strdup(&arg_root_shell, optarg);
+                        if (r < 0)
+                                return log_oom();
+
+                        break;
```
Comment (poettering) [inline src/firstboot/firstboot.c:1178]: validate with `valid_shell()` here too?

File: src/firstboot/firstboot.c [hunk 6]
```diff
@@ -748,13 +788,36 @@ static int process_root_password(void) {
...
 
-        if (arg_delete_root_password) {
+        if (arg_copy_root_shell && arg_root) {
+                struct passwd *p;
+
+                errno = 0;
+                p = getpwnam("root");
+                if (!p && errno != ENOENT) {
+                        if (!errno)
+                                errno = EIO;
```
Comment (poettering) [inline src/firstboot/firstboot.c]: errno_or_else()…

File: src/firstboot/firstboot.c (hunk 3 above)
Comment (poettering) [reply src/firstboot/firstboot.c]: so, getpwnam() is not great at returning useful error codes. Could be ESRCH too. The NSS modules backing it like to return any error code they like...

File: src/firstboot/firstboot.c [hunk 7]
```diff
@@ -774,93 +777,68 @@ static int write_root_shadow(const char *shadow_path, const char *hashed_passwor
...
+        const char *etc_passwd, *etc_shadow;
         int r;
 
+        etc_passwd = prefix_roota(arg_root, "/etc/passwd");
         etc_shadow = prefix_roota(arg_root, "/etc/shadow");
-        if (laccess(etc_shadow, F_OK) >= 0 && !arg_force)
+
+        /* Reduce the number of possible scenarios by always making sure we have permission to modify both
+         * /etc/passwd and /etc/shadow. */
+        if ((laccess(etc_passwd, F_OK) >= 0 || laccess(etc_shadow, F_OK) >= 0) && !arg_force)
```
Comment (poettering) [inline src/firstboot/firstboot.c:790]: this is not a permssion check. It's an existance check, i.e. F_OK

File: src/firstboot/firstboot.c [hunk 8]
```diff
@@ -774,93 +777,68 @@ static int write_root_shadow(const char *shadow_path, const char *hashed_passwor
...
-
-                log_info("%s written", etc_passwd);
-
-                return 0;
-        }
-
         if (arg_copy_root_password && arg_root) {
                 struct spwd *p;
 
-                errno = 0;
```
Comment (poettering) [inline src/firstboot/firstboot.c:769]: this should stay. errno handling is notoriously unreliable with password handling function (NSS modules) and documentation vague

File: src/firstboot/firstboot.c (hunk 7 above)
Comment (daandemeyer) [reply src/firstboot/firstboot.c:790]: My reasoning was that we only have permission from the user to modify these files if they don't exist yet or --force is specified, I'll update the comment to something more clarifying

File: src/firstboot/firstboot.c (hunk 8 above)
Comment (daandemeyer) [reply src/firstboot/firstboot.c:769]: I should have read the manpage.

**Comment by DaanDeMeyer** [comment]
Addressed the comments and switched the order of the two commits. First comes the refactoring, then the new feature. Should be a lot more logical to follow.

**Comment by poettering** [comment]
man page still missing, see earlier review

**Comment by poettering** [comment]
(looks good codewise otherwise)

**Comment by poettering** [comment]
btw, did you see my earlier comments that we probably should at least warn if the specified shell doesn't exist in the image? i.e. a quick call to chase_symlinks() with the specified shell and an error message if it cannot be found?

**Comment by DaanDeMeyer** [comment]
Yes I did, it was getting late yesterday so I didn't add it yet. Should I modify valid_shell to do the chase_symlinks call or will that break other stuff?

**Comment by poettering** [comment]
I'd keep it separate. valid_shell should not go to disk I think. it should be a superficial check just checking if the value makes any sense at all without matching it up with reality on some system.

**Pull request merged**